src/TypeAheadDropdown.jsx

```diff
@@ -0,0 +1,263 @@
+                    onClick={handleAddNewClick}
+                >
+                    <i className="fas fa-plus" /> Add new
+                </div>
+            </div>
+        );
+    }
+
+    return (
+        <div
+            className={classnames(styles.typeAheadRoot, "form-group", {
+                "invalid": props.invalid,
+            })}
+            data-testid="root"
+            onClick={(e) => e.stopPropagation()}
+        >
+            {props.label && (
+                <label data-testid="label">
+                    {props.label} {props.isRequired && <i className="fa fa-asterisk text-warn text-sm" /> }
+                </label>
+            )}
+            <div className={styles.inputContainer} ref={rootRef}>
+                <input
+                    value={currentSearchValue}
+                    data-testid="type-ahead-dropdown-input"
+                    onChange={(e) => handleSearch(e.target.value)}
+                    autoFocus={props.autoFocus}
+                    placeholder={props.placeholder}
+                    type={props.type}
+                    onFocus={handleFocus}
+                />
+                <button
+                    className={styles.selectButton}
+                    onClick={handleToggleSuggestionList}
+                    type="button"
+                >
+                    <i className={classnames("fas fa-chevron-down", styles.selectButtonIcon, {[styles.expanded]: showSuggestionList})} />
+                </button>
+            </div>
+            {suggestionListContent && (
+                <div className={classnames(styles.suggestionList, {
+                    [styles.top]: showTop,
+                })}>
+                    {suggestionListContent}
+                </div>
+            )}
+        </div>
+    );
+}
+
+TypeAheadDropdown.propTypes = {
+    value: PropTypes.any,
+    onChange: PropTypes.func.isRequired,
+    label: PropTypes.string,
+    name: PropTypes.string,
+    invalid: PropTypes.bool,
+    isRequired: PropTypes.bool,
+    autoFocus: PropTypes.bool,
+    placeholder: PropTypes.string,
+    type: PropTypes.string,
+    renderSuggestion: PropTypes.func.isRequired,
+    onSuggestionSelected: PropTypes.func,
+    onSelectDisplayField: PropTypes.string,
+    onSelectReturnField: PropTypes.string,
+    onAddNewOption: PropTypes.shape({
+        create: PropTypes.func,
+        afterCreate: PropTypes.func,
+    }),
+    onSearch: PropTypes.oneOfType([
+        // For in-component-searching (options are already loaded)
+        PropTypes.shape({
+            options: PropTypes.array,
+            searchFields: PropTypes.oneOfType([
+                PropTypes.string,
+                PropTypes.arrayOf(PropTypes.string),
+            ]),
+        }),
+        // Can be a callback function (asynchronous api searching)
+        PropTypes.shape({
+            apiMethod: PropTypes.func,
+            searchFilters: PropTypes.object,
+        }),
+    ]).isRequired,
```

Comment:
Let's talk about the design of this interface. We have two options we have to cover here. We can do an `onSearch` which will make the component externally controlled or a `searchFields` which will make the component internally controlled. As the search text changes, we need to make the determination based on the available props if we should have externally or internally controlled component. If we have `onSearch` it should render the options provided every time without filtering. If there are `searchFields` it should limit the options internally based on the search string and the fields to search.
This is similar to a native `input` where it is either uncontrolled (allowing you to put in whatever) or controlled (has to have an `onChange`). While that puts a little bit more work on the rendering component to handle the search, it allows for more flexibility in the case that the rendering component may just want to filter props in memory but in a custom way.

Reply:
Another possibility is to give the component an API method and a way to render. I don't know if that makes sense under the `onSearch` grouping though. I think we can just flatten it to regular props.

Reply:
I think I am following. How simple should we make this component? The props are a bit confusing right now because this component is covering so many different use cases (internal search, async search, what field to show when selected, what field to return when selected, handling creating a new option if none are found, etc).

Perhaps we can break it down into a few components...
- `TypeAheadDropdown`
  - Just needs `onSearch`, `suggestions`, and `renderSuggestion` props
- `AsyncTypeAhead`
  - Wraps the TypeAheadDropdown and handles searching via an apiMethod
  - Can also handle "adding a new option" when no options are found (StaticTypeAhead wouldn't need this functionality)
- `StaticTypeAhead`
  - Wraps the TypeAheadDropdown and does the searching internally

So really we would only directly use the AsyncTypeAhead or the StaticTypeAhead depending on use case.